### PR TITLE
Converted MYSQL syntax to PostgreSQL

### LINKED ISSUE #1: Test with current version
Please test the PostgreSQL integration with the current version of UserSpice, and fix inconsistencies coming from the updated base version, if any.

Please use PRs from pushed feature branches from now on.

### PR DESCRIPTION
**Pull request by ssobii2**

### CODE REVIEW

**Comment by GaborFarkas** [comment]
Please change the commit message to something that makes sense.

File: install/install/includes/sql.sql [hunk 1]
```diff
@@ -628,7 +627,7 @@ CREATE TABLE users (
 --
 
 INSERT INTO users (id, permissions, email, email_new, username, password, pin, fname, lname, language, email_verified, vericode, vericode_expiry, oauth_provider, oauth_uid, gender, locale, gpluslink, account_owner, account_id, account_mgr, fb_uid, picture, created, protected, msg_exempt, dev_user, msg_notification, cloak_allowed, oauth_tos_accepted, un_changed, force_pr, logins, last_login, join_date, modified, active) VALUES
-(1, 1, 'userspicephp@userspice.com', NULL, 'admin', '$2y$12$1v06jm2KMOXuuo3qP7erTuTIJFOnzhpds1Moa8BadnUUeX0RV3ex.', NULL, 'The', 'Admin', 'en-US', true, 'nlPsJDtyeqFWsS', NULL, '', '', '', '', '', 1, 0, 0, '', '', '0000-00-00 00:00:00', true, true, false, true, true, NULL, false, false, 0, '2022-12-23 07:16:27', '2022-12-25 00:00:00', '2016-01-01 00:00:00', true);
```
Comment (GaborFarkas) [inline install/install/includes/sql.sql:631]: Is there a problem with `0000-00-00 00:00:00` in postgres?

File: install/install/includes/sql.sql [hunk 2]
```diff
@@ -869,53 +868,53 @@ INSERT INTO us_management (id, page, view, feature, access) VALUES
...
+  disabled smallint DEFAULT 0,
+  justify varchar(10) DEFAULT 'right'
+);
 
 --
 -- Dumping data for table `us_menus`
 --
 
-INSERT INTO `us_menus` (`id`, `menu_name`, `type`, `nav_class`, `theme`, `z_index`, `brand_html`, `disabled`, `justify`) VALUES
-(1, 'Main Menu', 'horizontal', '', 'dark', 50, '&lt;a href=&quot;{{root}}&quot; &gt;\r\n&lt;img src=&quot;{{root}}users/images/logo.png&quot; /&gt;', 0, 'right'),
```
Comment (GaborFarkas) [inline install/install/includes/sql.sql:889]: Please revert this back to the HTML-escaped format, if possible.

File: install/install/includes/sql.sql [hunk 3]
```diff
@@ -869,53 +868,53 @@ INSERT INTO us_management (id, page, view, feature, access) VALUES
...
-(2, 'Dashboard Menu', 'horizontal', NULL, 'dark', 55, '&lt;a href=&quot;{{root}}&quot; title=&quot;Home Page&quot;&gt;\r\n&lt;img src=&quot;{{root}}users/images/logo.png&quot; alt=&quot;Main logo&quot; /&gt;&lt;/a&gt;', 0, 'right');
```
Comment (GaborFarkas) [inline install/install/includes/sql.sql:890]: Please revert this back to the HTML-escaped format, if possible.

File: users/classes/Menu.php
Comment (GaborFarkas) [inline users/classes/Menu.php:1]: Please undo these changes and make sure that the original logic can work with the postgres connection.

File: users/updates/components/0DaShInkjV86.php [hunk 4]
```diff
@@ -33,11 +33,6 @@
 (14, '_admin_user.php', 'user', 'User Management', ''),
 (15, '_admin_users.php', 'users', 'User Management', '')");
 
-$db->query("ALTER TABLE `us_management`
-  ADD PRIMARY KEY (`id`)");
-
-$db->query("ALTER TABLE `us_management`
-  MODIFY `id` int(11) NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=16;
-COMMIT");
+$db->query("ALTER SEQUENCE us_management_id_seq RESTART WITH 16;");
```
Comment (GaborFarkas) [inline users/updates/components/0DaShInkjV86.php]: Is this line needed? You defined `id` as `PK` in the table definition, so the sequence should automatically increment on every inserted row.

File: users/updates/components/2022-11-06a.php [hunk 5]
```diff
@@ -5,39 +5,35 @@
...
+    type varchar(75),
+    nav_class varchar(255),
+    theme varchar(25),
+    z_index int,
+    brand_html text,
+    disabled smallint DEFAULT 0
+  )");
+
+$db->query("INSERT INTO us_menus (id, menu_name, type, nav_class, theme, z_index, brand_html, disabled)
+            VALUES (1, 'Main Menu', 'horizontal', '', 'dark', 50, '<a href=\"{{root}}\"> <img src=\"{{root}}users/images/logo.png\" /></a>', 0)");
```
Comment (GaborFarkas) [inline users/updates/components/2022-11-06a.php]: Please change back special characters to HTML-encoded ones here as well.

File: users/updates/components/69qa8h6E1bzG.php [hunk 6]
```diff
@@ -5,7 +5,7 @@
 //Release Date Unknown
 //Rewrote 2019-02-19 BA
 
-$db->query("UPDATE logs SET logtype = ? WHERE logtype = ? AND lognote LIKE ?",array("IP Logging","User","%blacklisted%attempted%visit"));
```
Comment (GaborFarkas) [inline users/updates/components/69qa8h6E1bzG.php]: What was the problem with this syntax?

File: users/updates/components/86FkFVV4TGRg.php [hunk 7]
```diff
@@ -7,9 +7,9 @@
 $countE=0;
 
 $db->query("ALTER TABLE groups_menus
-  MODIFY COLUMN id int(11) UNSIGNED NOT NULL AUTO_INCREMENT,
-  MODIFY COLUMN group_id int(11) UNSIGNED NOT NULL,
-  MODIFY COLUMN menu_id int(11) UNSIGNED NOT NULL");
+  ALTER COLUMN id SET DATA TYPE SERIAL,
+  ALTER COLUMN group_id SET DATA TYPE INTEGER,
```
Comment (GaborFarkas) [inline users/updates/components/86FkFVV4TGRg.php:14]: Please add `NOT NULL` constraint as well, if not already present.

File: users/updates/components/86FkFVV4TGRg.php [hunk 8]
```diff
@@ -7,9 +7,9 @@
...
+  ALTER COLUMN menu_id SET DATA TYPE INTEGER");
```
Comment (GaborFarkas) [inline users/updates/components/86FkFVV4TGRg.php]: Please add `NOT NULL` constraint as well, if not already present.

File: users/updates/components/VLBp32gTWvEo.php [hunk 9]
```diff
@@ -7,7 +7,7 @@
 
 $countE=0;
 
-$db->query("ALTER TABLE users ADD COLUMN vericode_expiry timestamp AFTER `vericode`");
+$db->query("ALTER TABLE users ADD COLUMN vericode_expiry timestamp AFTER vericode");
```
Comment (GaborFarkas) [inline users/updates/components/VLBp32gTWvEo.php]: Is there an `AFTER` keyword in PostgreSQL?

File: users/updates/components/c7tZQf926zKq.php [hunk 10]
```diff
@@ -35,11 +35,11 @@
 }
 
 $db->query("CREATE TABLE us_fingerprint_assets (
-  kFingerprintAssetID int(11) UNSIGNED NOT NULL PRIMARY KEY AUTO_INCREMENT,
-  fkFingerprintID int(11) NOT NULL,
-  IP_Address varchar(255) NOT NULL,
-  User_Browser varchar(255) NOT NULL,
-  User_OS varchar(255) NOT NULL
+  kFingerprintAssetID SERIAL PRIMARY KEY,
+  fkFingerprintID INT NOT NULL,
```
Comment (GaborFarkas) [inline users/updates/components/c7tZQf926zKq.php:39]: Please remain syntactically consistent. If you are using `integer`, use it everywhere.

File: users/views/_admin_menu_edit.php [hunk 11]
```diff
@@ -14,7 +14,7 @@
 foreach ($index as $i) {
   $indexes[] = $i->z_index;
 }
-$parentOpts = $db->query("SELECT * FROM us_menu_items WHERE menu = ? AND `type` = ? ORDER BY label", [$menuId, "dropdown"])->results();
+$parentOpts = $db->query("SELECT * FROM us_menu_items WHERE menu = $1 AND \"type\" = $2 ORDER BY label", [$menuId, "dropdown"])->results();
```
Comment (GaborFarkas) [inline users/views/_admin_menu_edit.php]: What is the problem with positional `?` placeholders in a PDO request with PostgreSQL?

File: users/views/_admin_menu_edit.php [hunk 12]
```diff
@@ -282,10 +282,11 @@
             </div>
 
             <div class="form-check">
-              <input class="form-check-input" type="checkbox" value="1" id="disabled" name="disabled" <?= $item->disabled ? "checked" : "" ?> onchange="setDirty(true)">
+              <input class="form-check-input" type="checkbox" value="1" id="disabled" name="disabled" <?= $item->disabled ? "checked" : 0 ?> onchange="setDirty(true)">
```
Comment (GaborFarkas) [inline users/views/_admin_menu_edit.php]: Please revert this change.

File: users/views/_admin_menu_edit.php [hunk 13]
```diff
@@ -282,10 +282,11 @@
...
               <label class="form-check-label" for="disabled">
                 Disabled
               </label>
+              <input type="hidden" name="disabled" value="0">
```
Comment (GaborFarkas) [inline users/views/_admin_menu_edit.php]: Please remove this change.

File: users/views/_admin_menu_edit.php [hunk 14]
```diff
@@ -363,10 +364,11 @@
             </div>
 
             <div class="form-check">
-              <input class="form-check-input" type="checkbox" value="1" id="disabled" name="disabled" <?= $menu->disabled ? "checked" : "" ?> onchange="setDirty(true)">
+              <input class="form-check-input" type="checkbox" value="1" id="disabled" name="disabled" <?= $menu->disabled ? "checked" : 0 ?> onchange="setDirty(true)">
```
Comment (GaborFarkas) [inline users/views/_admin_menu_edit.php]: Please revert this change.

File: users/views/_admin_menu_edit.php [hunk 15]
```diff
@@ -363,10 +364,11 @@
...
               <label class="form-check-label" for="disabled">
                 Disabled
               </label>
+              <input type="hidden" name="disabled" value="0">
```
Comment (GaborFarkas) [inline users/views/_admin_menu_edit.php]: Please remove this change.

File: users/views/_admin_menus.php [hunk 16]
```diff
@@ -1,7 +1,7 @@
 <?php
 $db = DB::getInstance();
 $menus = $db->query("SELECT * FROM us_menus")->results();
-$count = $db->query("SELECT z_index as z, COUNT(*) z_index FROM us_menus GROUP BY z_index HAVING z_index > 1")->results();
+$count = $db->query("SELECT z_index, COUNT(*) as count FROM us_menus GROUP BY z_index HAVING z_index > 1")->results();
```
Comment (GaborFarkas) [inline users/views/_admin_menus.php]: I don't think this query does the same thing as the original one. Maybe `SELECT z_index as z, COUNT(*) as z_index FROM us_menus GROUP BY z_index HAVING z_index > 1`.

**Review by GaborFarkas — CHANGES_REQUESTED** [review]

File: install/install/includes/sql.sql (hunk 1 above)
Comment (ssobii2) [reply install/install/includes/sql.sql:631]: Yes, professor timestamp doesn't accept this in PostgreSQL. I found this solution online. Should I go with this?

image

File: install/install/includes/sql.sql (hunk 2 above)
Comment (ssobii2) [reply install/install/includes/sql.sql:889]: All the changes you reviewed, so how can I handle it? The changes you accept and changes you reject. Do I need to make those changes again in VS code and then push again? Is this how it works? I have never done this before, so I don't want to mess this up.

Comment (ssobii2) [reply install/install/includes/sql.sql:889]: Ok, Professor, I will revert it to the original. And yes, this looks wrong. For some stuff, I used AI to convert. Next time I will be more careful.

File: install/install/includes/sql.sql (hunk 1 above)
Comment (GaborFarkas) [reply install/install/includes/sql.sql:631]: Let's time it before the modified timestamp (2016-01-01 00:00:00) to make sure we won't bump into any exceptions from the modified < created situation. Use 2015-01-01 00:00:00, please.

File: users/classes/Menu.php
Comment (ssobii2) [reply users/classes/Menu.php:1]: It works with the old classic menu. I will look into it.

File: users/updates/components/0DaShInkjV86.php (hunk 4 above)
Comment (ssobii2) [reply users/updates/components/0DaShInkjV86.php]: I made the changes according to the MySQL code; as you can see, AUTO_INCREMENT is set to 16. I don't know if I need to keep it, but I had a similar PK conflict somewhere else before where the sequence started from 1, and the record was not inserted, so this manually starts it from 16. What do you think?

File: install/install/includes/sql.sql (hunk 2 above)
Comment (GaborFarkas) [reply install/install/includes/sql.sql:889]: Please NEVER do that in the future.

File: users/updates/components/86FkFVV4TGRg.php (hunk 7 above)
Comment (ssobii2) [reply users/updates/components/86FkFVV4TGRg.php:14]: I must have forgotten I added to some other ones I missed too.

File: users/updates/components/VLBp32gTWvEo.php (hunk 9 above)
Comment (ssobii2) [reply users/updates/components/VLBp32gTWvEo.php]: So I was confused with this AI suggested there is no AFTER keyword one time. I even googled it, but I am unsure if this keyword can be used. I found an example of it being used with a TRIGGER. Also, the keyword shows here on the Postgre website.

https://www.postgresql.org/docs/current/sql-keywords-appendix.html

It shows as non-reserved, which means it has context in some areas but not all. What do you this, professor? I will test it, too and let you know,

File: users/updates/components/c7tZQf926zKq.php (hunk 10 above)
Comment (ssobii2) [reply users/updates/components/c7tZQf926zKq.php:39]: Should I change all of these later when changing Boolean to smallint also?

File: users/views/_admin_menu_edit.php (hunk 11 above)
Comment (ssobii2) [reply users/views/_admin_menu_edit.php]: So both work, but I had a SQL error when using the ? Placeholder. The AI suggested using the $ placeholder, so the error was gone. The query was also working. Can it be a different problem? Should I keep it or revert it?

File: users/views/_admin_menus.php (hunk 16 above)
Comment (ssobii2) [reply users/views/_admin_menus.php]: I checked with chat gpt, and yours and mine are doing the same thing. Just the alias is different. I am using count, and yours is using z. I don't know if AI gave me the correct answer. Anyway, I will use your professor.

**Comment by ssobii2** [comment]
Hi Professor, after making some changes and undoing the changes, something weird is happening in my Linux. Now the database is not updating like the updates are not applying. What can be the issue? It also shows that your database needs to be updated; below, it doesn't even display the database name it showed before.
I checked the Apache logs, too, for the updated file. It's saying some write issue to a file. I tried the try-catch block, and the file cannot be opened. Another issue with plugins is that when I open the page, it says the file is not writable. I have the same permissions as before. I don't know why this is happening.
Should I push the changes? It's still working without the updates.

**Comment by ssobii2** [comment]
image

File: users/updates/components/VLBp32gTWvEo.php (hunk 9 above)
Comment (ssobii2) [reply users/updates/components/VLBp32gTWvEo.php]: Ok, so it already exists in the sql.sql file. I don't know how else to test it. Shouldn't we take dump of the database before the installation, like when the updates are not applied, so that these updates can be applied later when the user goes to /users/updates or click the link on the dashboard?

File: users/updates/components/0DaShInkjV86.php (hunk 4 above)
Comment (GaborFarkas) [reply users/updates/components/0DaShInkjV86.php]: I think this particular case should work without the `ALTER SEQUENCE` statement, as the `id` is `PK` in the table definition. Please try it out without that line.

File: users/updates/components/VLBp32gTWvEo.php (hunk 9 above)
Comment (GaborFarkas) [reply users/updates/components/VLBp32gTWvEo.php]: The `AFTER` keyword is used in PostgreSQL's [trigger system](https://www.postgresql.org/docs/current/plpgsql-trigger.html). It cannot be used for specifying the position of a newly inserted column. Please remove and check for other occurrences.

File: users/views/_admin_menu_edit.php (hunk 11 above)
Comment (GaborFarkas) [reply users/views/_admin_menu_edit.php]: Please revert.

Comment (ssobii2) [reply users/views/_admin_menu_edit.php]: I reverted other placeholders too in this file

File: install/install/includes/sql.sql [hunk 17]
```diff
@@ -405,7 +405,7 @@ INSERT INTO profiles (id, user_id, bio) VALUES
 --
 
 CREATE TABLE settings (
-  id serial NOT NULL,
+  id serial PRIMARY KEY NOT NULL,
```
Comment (GaborFarkas) [inline install/install/includes/sql.sql]: NN is not needed for PK

File: install/install/includes/sql.sql (hunk 1 above)
Comment (GaborFarkas) [reply install/install/includes/sql.sql:631]: Please change as requested.

Comment (GaborFarkas) [reply install/install/includes/sql.sql:631]: Sorry, I did not see your most recent changes.

File: index.php
Comment (GaborFarkas) [inline index.php:1]: Please revert these changes and untrack this file in your local repository.

File: users/classes/Token.future
Comment (GaborFarkas) [inline users/classes/Token.future:1]: Please restore this file.

File: users/classes/Token.old
Comment (GaborFarkas) [inline users/classes/Token.old:1]: Please restore this file.

File: users/updates/components/86FkFVV4TGRg.php
Comment (GaborFarkas) [inline users/updates/components/86FkFVV4TGRg.php:1]: Have you tested this? I highly doubt `SET DATA TYPE` accepts `NOT NULL`.

File: usersc/includes/dashboard_overrides.php
Comment (GaborFarkas) [inline usersc/includes/dashboard_overrides.php:1]: Please do not commit your local configurations. Untrack the file if you cannot remember what you should not commit.

**Review by GaborFarkas — CHANGES_REQUESTED** [review]

File: users/classes/Token.future
Comment (ssobii2) [reply users/classes/Token.future:1]: it automatically got deleted when I went to users/updates I don't know but I will restore both of them

File: index.php
Comment (ssobii2) [reply index.php:1]: The same way i did with the init.php right?

File: users/updates/components/86FkFVV4TGRg.php
Comment (ssobii2) [reply users/updates/components/86FkFVV4TGRg.php:1]: I tested it, and it doesn't work, but the solution is slightly different, which I implemented. Please feel free to check.
Also, should I test other updates? most will work, but testing will take some time.

File: usersc/includes/dashboard_overrides.php
Comment (ssobii2) [reply usersc/includes/dashboard_overrides.php:1]: Please let me know which files I should untrack. In the file you mentioned here, I didn't make any changes; it automatically did. I will see from now on if extra files are added to git. I will untrack them.

File: index.php
Comment (GaborFarkas) [reply index.php:1]: Yes.

File: usersc/includes/dashboard_overrides.php
Comment (GaborFarkas) [reply usersc/includes/dashboard_overrides.php:1]: If you see some files automatically changed, either discard the changes, untrack the files, or just leave them out from your commits. It's up to you to decide, but I don't want to see any of those changes in the repo.

**Review by GaborFarkas — CHANGES_REQUESTED** [review]
As a final step, please do an interactive rebase and rename your commits to have more meaningful messages. When done, please force-push the postgres branch.

**Comment by ssobii2** [comment]
> As a final step, please do an interactive rebase and rename your commits to have more meaningful messages. When done, please force-push the Postgres branch.

So I was using `#3` for the latest commit, and it got the original forked repo owner name, so I changed it, but how can I do it for the previous commits? Also, any suggestions for naming, as It just involves fixing and updating them with your changes?

**Comment by ssobii2** [comment]
> > As a final step, please do an interactive rebase and rename your commits to have more meaningful messages. When done, please force-push the Postgres branch.
> 
> So I was using `#3` for the latest commit, and it got the original forked repo owner name, so I changed it, but how can I do it for the previous commits? Also, any suggestions for naming, as It just involves fixing and updating them with your changes?

It did it again you can see the mudmin part

**Comment by ssobii2** [comment]
Is this the way to do it? I am asking because I don't want to mess this up. It's better to ask, in my opinion.

image

And yes, I am not using AI intentionally. It just opens at the side and copies the answer for the references below, which I double-checked.

**Comment by GaborFarkas** [comment]
> Is this the way to do it? I am asking because I don't want to mess this up. It's better to ask, in my opinion.
> 
> image
> 
> And yes, I am not using AI intentionally. It just opens at the side and copies the answer for the references below, which I double-checked.

Yes, `git rebase -i HEAD~<N>` is one way to do it. Please do not use hashtags in commit messages, as they are used as pointers by GitHub.

**Comment by ssobii2** [comment]
Hi Professor,

So I was having issues with the git push --force after modifying the commit messages. Some problem with git upstream doesn't exist for the postgres branch. I tried to fix it, and it pushed the changes to the branch without you merging it. I changed nothing, just the commit messages, so you can still check and merge if necessary. After you are done, I will clone the branch again so I don't have further issues and start fresh for the next task.

**Comment by ssobii2** [comment]
Another thing is how can I allow git only to push to pull request and not the main branch because you can see when I force pushed it pushed to the branch also.

**Review by GaborFarkas — APPROVED** [review]
